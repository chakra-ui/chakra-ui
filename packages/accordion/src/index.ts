export * from "./accordion"
export * from "./use-accordion"
