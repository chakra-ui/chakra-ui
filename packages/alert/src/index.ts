export { Alert, type AlertProps } from "./alert"
export { useAlertStyles, type AlertStatus } from "./alert-context"
export {
  AlertDescription,
  type AlertDescriptionProps,
} from "./alert-description"
export { AlertIcon, type AlertIconProps } from "./alert-icon"
export { AlertTitle, type AlertTitleProps } from "./alert-title"
