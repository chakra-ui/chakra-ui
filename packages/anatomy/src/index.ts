import { anatomy } from "@chakra-ui/theme-tools"

/**
 * **Accordion anatomy**
 * - Root: the root container of the accordion
 * - Container: the accordion item contains the button and panel
 * - Button: the button is the trigger for the panel
 * - Panel: the panel is the content of the accordion item
 * - Icon: the expanded/collapsed icon
 */
export const accordionAnatomy = anatomy("accordion")
  .parts("root", "container", "button", "panel")
  .extend("icon")

/**
 * **Alert anatomy**
 * - Title: the alert's title
 * - Description: the alert's description
 * - Icon: the alert's icon
 */
export const alertAnatomy = anatomy("alert")
  .parts("title", "description", "container")
  .extend("icon")

/**
 * **Avatar anatomy**
 * - Container: the container for the avatar
 * - Label: the avatar initials text
 * - Excess Label: the label or text that represents excess avatar count.
 * Typically used in avatar groups.
 * - Group: the container for the avatar group
 */
export const avatarAnatomy = anatomy("avatar")
  .parts("label", "badge", "container")
  .extend("excessLabel", "group")

/**
 * **Breadcrumb anatomy**
 * - Item: the container for a breadcrumb item
 * - Link: the element that represents the breadcrumb link
 * - Container: the container for the breadcrumb items
 * - Separator: the separator between breadcrumb items
 */
export const breadcrumbAnatomy = anatomy("breadcrumb")
  .parts("link", "item", "container")
  .extend("separator")

export const buttonAnatomy = anatomy("button").parts()

export const checkboxAnatomy = anatomy("checkbox")
  .parts("control", "icon", "container")
  .extend("label")

export const circularProgressAnatomy = anatomy("progress")
  .parts("track", "filledTrack")
  .extend("label")

export const drawerAnatomy = anatomy("drawer")
  .parts("overlay", "dialogContainer", "dialog")
  .extend("header", "closeButton", "body", "footer")

export const editableAnatomy = anatomy("editable").parts(
  "preview",
  "input",
  "textarea",
)

export const formAnatomy = anatomy("form").parts(
  "container",
  "requiredIndicator",
  "helperText",
)

export const formErrorAnatomy = anatomy("formError").parts("text", "icon")

export const inputAnatomy = anatomy("input").parts("addon", "field", "element")

export const listAnatomy = anatomy("list").parts("container", "item", "icon")

export const menuAnatomy = anatomy("menu")
  .parts("button", "list", "item")
  .extend("groupTitle", "command", "divider")

export const modalAnatomy = anatomy("modal")
  .parts("overlay", "dialogContainer", "dialog")
  .extend("header", "closeButton", "body", "footer")

export const numberInputAnatomy = anatomy("numberinput").parts(
  "root",
  "field",
  "stepperGroup",
  "stepper",
)

export const pinInputAnatomy = anatomy("pininput").parts("field")

export const popoverAnatomy = anatomy("popover")
  .parts("content", "header", "body", "footer")
  .extend("popper", "arrow", "closeButton")

export const progressAnatomy = anatomy("progress").parts(
  "label",
  "filledTrack",
  "track",
)

export const radioAnatomy = anatomy("radio").parts(
  "container",
  "control",
  "label",
)

export const selectAnatomy = anatomy("select").parts("field", "icon")

export const sliderAnatomy = anatomy("slider").parts(
  "container",
  "track",
  "thumb",
  "filledTrack",
)

export const statAnatomy = anatomy("stat").parts(
  "container",
  "label",
  "helpText",
  "number",
  "icon",
)

export const switchAnatomy = anatomy("switch").parts(
  "container",
  "track",
  "thumb",
)

export const tableAnatomy = anatomy("table").parts(
  "table",
  "thead",
  "tbody",
  "tr",
  "th",
  "td",
  "tfoot",
  "caption",
)

export const tabsAnatomy = anatomy("tabs").parts(
  "root",
  "tab",
  "tablist",
  "tabpanel",
  "tabpanels",
  "indicator",
)

/**
 * **Tag anatomy**
 * - Container: the container for the tag
 * - Label: the text content of the tag
 * - closeButton: the close button for the tag
 */
export const tagAnatomy = anatomy("tag").parts(
  "container",
  "label",
  "closeButton",
)
