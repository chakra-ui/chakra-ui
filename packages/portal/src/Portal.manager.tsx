import * as React from "react"
import { createContext } from "@chakra-ui/utils"
import { useSafeLayoutEffect, useForceUpdate } from "@chakra-ui/hooks"

export interface PortalManagerContext {
  node: HTMLElement
  zIndex?: number
}

const [PortalManagerProvider, usePortalManager] = createContext<
  PortalManagerContext
>({
  strict: false,
})

export { usePortalManager }

export interface PortalManagerProps {
  /**
   * Child elements of the Portal manager
   * Ideally, it should be at the top-level
   * of your application
   */
  children?: React.ReactNode
  /**
   * [Z-Index war] If your has multiple elements
   * with z-index clashing, you might need to
   * apply a z-index to the Portal manager
   */
  zIndex?: number
}

/**
 * Manage multiple portals within an application
 * Inspired by BaseWeb's LayerManager component
 */
export function PortalManager(props: PortalManagerProps) {
  const { children, zIndex } = props

  // The element that wraps the stacked layers
  const ref = React.useRef<HTMLDivElement>(null)

  const forceUpdate = useForceUpdate()

  // force an update on mount so the Provider works correctly
  useSafeLayoutEffect(() => {
    forceUpdate()
  }, [])

  // let's detect if use has mutiple instances of this component
  const parentManager = usePortalManager()

  const context = {
    node: parentManager?.node || ref.current,
    zIndex,
  }

  return (
    <PortalManagerProvider value={context}>
      {children}
      <div className="chakra-portal-manager" ref={ref} />
    </PortalManagerProvider>
  )
}
