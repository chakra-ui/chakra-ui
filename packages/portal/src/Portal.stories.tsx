import * as React from "react"
import { Portal, PortalManager } from "."

export default {
  title: "Portal",
  decorators: [
    (Story: Function) => (
      <PortalManager>
        <Story />
      </PortalManager>
    ),
  ],
}

export const BasicPortal = () => (
  <>
    <p>Welcome</p>
    <Portal>This text has been portaled</Portal>
  </>
)

export const WithMountRef = () => {
  const ref = React.useRef<HTMLDivElement>(null)

  return (
    <>
      <p>Welcome</p>
      <Portal container={() => ref.current}>
        <span>This text has been portaled</span>
      </Portal>
      <div id="iframe" ref={ref}>
        Portal Div
      </div>
    </>
  )
}

function Wrapper(props: any) {
  const { offset, color, children } = props
  return (
    <div
      style={{
        position: "fixed",
        top: offset || "46%",
        left: offset || "46%",
        width: "200px",
        paddingTop: "20px",
        paddingBottom: "20px",
        paddingLeft: "20px",
        paddingRight: "20px",
        backgroundColor: color,
        textAlign: "center",
      }}
    >
      {children}
    </div>
  )
}

export const NestedPortals = () => {
  return (
    <Portal>
      <Wrapper color="red">Welcome</Wrapper>
      <Portal>
        <Wrapper offset="40%" color="green">
          Welcome
        </Wrapper>
        <Portal>
          <Wrapper offset="30%" color="tomato">
            Welcome
          </Wrapper>
        </Portal>
      </Portal>
    </Portal>
  )
}
