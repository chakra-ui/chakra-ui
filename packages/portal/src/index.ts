export * from "./portal-manager"
export * from "./portal"
