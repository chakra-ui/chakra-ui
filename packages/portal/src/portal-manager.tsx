import { __DEV__ } from "@chakra-ui/utils"
import { createContext } from "@chakra-ui/react-utils"
import * as React from "react"

interface PortalManagerContext {
  zIndex?: number
}

const [
  PortalManagerContextProvider,
  usePortalManager,
] = createContext<PortalManagerContext | null>({
  strict: false,
  name: "PortalManagerContext",
})

export { usePortalManager }

export interface PortalManagerProps {
  children?: React.ReactNode
  /**
   * [Z-Index war] If your has multiple elements
   * with z-index clashing, you might need to apply a z-index to the Portal manager
   */
  zIndex?: number
}

export function PortalManager(props: PortalManagerProps) {
  const { children, zIndex } = props
  return (
    <PortalManagerContextProvider value={{ zIndex }}>
      {children}
    </PortalManagerContextProvider>
  )
}

if (__DEV__) {
  PortalManager.displayName = "PortalManager"
}
