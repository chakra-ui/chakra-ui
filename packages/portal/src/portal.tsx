import { useForceUpdate, useSafeLayoutEffect } from "@chakra-ui/hooks"
import { isBrowser, __DEV__ } from "@chakra-ui/utils"
import { createContext } from "@chakra-ui/react-utils"
import * as React from "react"
import { createPortal } from "react-dom"
import { usePortalManager } from "./portal-manager"

type PortalContext = HTMLDivElement | null

const [PortalContextProvider, usePortalContext] = createContext<PortalContext>({
  strict: false,
  name: "PortalContext",
})

const PORTAL_CLASSNAME = "chakra-portal"
const PORTAL_SELECTOR = `.chakra-portal`

const Container: React.FC<{ zIndex: number }> = (props) => (
  <div
    className="chakra-portal-zIndex"
    style={{
      position: "absolute",
      zIndex: props.zIndex,
      top: 0,
      left: 0,
      right: 0,
      // NB: Don't add `bottom: 0`, it makes the entire app unusable
      // @see https://github.com/chakra-ui/chakra-ui/issues/3201
    }}
  >
    {props.children}
  </div>
)

/**
 * Portal that uses `document.body` as container
 */
const DefaultPortal: React.FC<{ appendToParentPortal?: boolean }> = (props) => {
  const { appendToParentPortal, children } = props

  const tempNode = React.useRef<HTMLDivElement | null>(null)
  const portal = React.useRef<HTMLDivElement | null>(null)

  const forceUpdate = useForceUpdate()

  const parentPortal = usePortalContext()
  const manager = usePortalManager()

  useSafeLayoutEffect(() => {
    if (!tempNode.current) return

    const doc = tempNode.current!.ownerDocument
    const host = appendToParentPortal ? parentPortal ?? doc.body : doc.body

    if (!host) return

    portal.current = doc.createElement("div")
    portal.current.className = PORTAL_CLASSNAME

    host.appendChild(portal.current)
    forceUpdate()

    const portalNode = portal.current
    return () => {
      if (host.contains(portalNode)) {
        host.removeChild(portalNode)
      }
    }
  }, [])

  const _children = manager?.zIndex ? (
    <Container zIndex={manager?.zIndex}>{children}</Container>
  ) : (
    children
  )

  return portal.current ? (
    createPortal(
      <PortalContextProvider value={portal.current}>
        {_children}
      </PortalContextProvider>,
      portal.current,
    )
  ) : (
    <span ref={tempNode} />
  )
}

interface ContainerPortalProps {
  containerRef: React.RefObject<HTMLElement | null>
  appendToParentPortal?: boolean
}

/**
 * Portal that uses a custom container
 */
const ContainerPortal: React.FC<ContainerPortalProps> = (props) => {
  const { children, containerRef, appendToParentPortal } = props
  const containerEl = containerRef.current
  const host = containerEl ?? (isBrowser ? document.body : undefined)

  const portal = React.useMemo(() => {
    const node = containerEl?.ownerDocument.createElement("div")
    if (node) node.className = PORTAL_CLASSNAME
    return node
  }, [containerEl])

  const forceUpdate = useForceUpdate()

  useSafeLayoutEffect(() => {
    forceUpdate()
  }, [])

  useSafeLayoutEffect(() => {
    if (!portal || !host) return
    host.appendChild(portal)
    return () => {
      host.removeChild(portal)
    }
  }, [portal, host])

  if (host && portal) {
    return createPortal(
      <PortalContextProvider value={appendToParentPortal ? portal : null}>
        {children}
      </PortalContextProvider>,
      portal,
    )
  }

  return null
}

export interface PortalProps {
  /**
   * The `ref` to the component where the portal will be attached to.
   */
  containerRef?: React.RefObject<HTMLElement | null>
  /**
   * The content or node you'll like to portal
   */
  children: React.ReactNode
  /**
   * If `true`, the portal will check if it is within a parent portal
   * and append itself to the parent's portal node.
   * This provides nesting for portals.
   *
   * If `false`, the portal will always append to `document.body`
   * regardless of nesting. It is used to opt out of portal nesting.
   */
  appendToParentPortal?: boolean
}

/**
 * Portal
 *
 * Declarative component used to render children into a DOM node
 * that exists outside the DOM hierarchy of the parent component.
 *
 * @see Docs https://chakra-ui.com/portal
 */

export function Portal(props: PortalProps) {
  const { containerRef, ...rest } = props
  return containerRef ? (
    <ContainerPortal containerRef={containerRef} {...rest} />
  ) : (
    <DefaultPortal {...rest} />
  )
}

Portal.defaultProps = {
  appendToParentPortal: true,
}

Portal.className = PORTAL_CLASSNAME
Portal.selector = PORTAL_SELECTOR

if (__DEV__) {
  Portal.displayName = "Portal"
}
