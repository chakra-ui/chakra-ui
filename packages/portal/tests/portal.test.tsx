import { render, screen } from "@chakra-ui/test-utils"
import * as React from "react"
import { PortalManager, Portal } from "../src"

test("should render portal", () => {
  const tools = render(
    <PortalManager>
      <Portal>This is a portal</Portal>
    </PortalManager>,
  )

  expect(tools.baseElement.innerHTML).toMatchSnapshot()
})

test("should render nested portal", () => {
  const tools = render(
    <PortalManager>
      <Portal>
        This is a portal.
        <Portal>This is a nested portal</Portal>
      </Portal>
    </PortalManager>,
  )

  expect(tools.asFragment()).toMatchSnapshot()

  const portals = Array.from(
    tools.baseElement.querySelectorAll(".chakra-portal"),
  )

  const [parentPortal, childPortal] = portals
  expect(parentPortal).toContainElement(childPortal as HTMLElement)
})

test("should render in a different node", () => {
  const tools = render(
    <PortalManager>
      <div data-testid="parent">
        <h1 data-testid="child-1">Foo</h1>
        <Portal>
          <h1 data-testid="child-2">Foo</h1>
        </Portal>
      </div>
    </PortalManager>,
  )

  expect(tools.asFragment()).toMatchSnapshot()

  const parent = screen.getByTestId("parent")

  const child1 = screen.getByTestId("child-1")
  const child2 = screen.getByTestId("child-2")

  expect(parent).toContainElement(child1)
  expect(parent).not.toContainElement(child2)
})

test("should render into a custom container", () => {
  const Custom = () => {
    const ref = React.useRef<any>(null)
    return (
      <PortalManager>
        <div data-testid="container" ref={ref} />
        <Portal getContainer={() => ref.current}>
          <h1 data-testid="heading">Hello world</h1>
        </Portal>
      </PortalManager>
    )
  }

  const tools = render(<Custom />)

  expect(tools.asFragment()).toMatchSnapshot()

  const heading = tools.getByTestId("heading")
  const container = tools.getByTestId("container")

  expect(container).toContainElement(heading)
})
