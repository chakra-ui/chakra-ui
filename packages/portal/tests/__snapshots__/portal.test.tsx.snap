// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render in a different node 1`] = `
<body
  class="chakra-ui-light"
>
  <div
    id="chakra-toast-portal"
  >
    <span
      id="chakra-toast-manager-top"
      style="position: fixed; z-index: 5500; pointer-events: none; margin: 0px auto; text-align: center; top: 0px; right: 0px; left: 0px;"
    />
    <span
      id="chakra-toast-manager-top-left"
      style="position: fixed; z-index: 5500; pointer-events: none; top: 0px; left: 0px;"
    />
    <span
      id="chakra-toast-manager-top-right"
      style="position: fixed; z-index: 5500; pointer-events: none; top: 0px; right: 0px;"
    />
    <span
      id="chakra-toast-manager-bottom-left"
      style="position: fixed; z-index: 5500; pointer-events: none; bottom: 0px; left: 0px;"
    />
    <span
      id="chakra-toast-manager-bottom"
      style="position: fixed; z-index: 5500; pointer-events: none; margin: 0px auto; text-align: center; bottom: 0px; right: 0px; left: 0px;"
    />
    <span
      id="chakra-toast-manager-bottom-right"
      style="position: fixed; z-index: 5500; pointer-events: none; bottom: 0px; right: 0px;"
    />
  </div>
  <div>
    <div
      data-testid="parent"
    >
      <h1
        data-testid="child-1"
      >
        Foo
      </h1>
    </div>
    <div
      class="chakra-portal-manager"
    >
      <div
        class="chakra-portal"
      >
        <h1
          data-testid="child-2"
        >
          Foo
        </h1>
      </div>
    </div>
  </div>
</body>
`;

exports[`should render into a custom container 1`] = `
<body
  class="chakra-ui-light"
>
  <div
    id="chakra-toast-portal"
  >
    <span
      id="chakra-toast-manager-top"
      style="position: fixed; z-index: 5500; pointer-events: none; margin: 0px auto; text-align: center; top: 0px; right: 0px; left: 0px;"
    />
    <span
      id="chakra-toast-manager-top-left"
      style="position: fixed; z-index: 5500; pointer-events: none; top: 0px; left: 0px;"
    />
    <span
      id="chakra-toast-manager-top-right"
      style="position: fixed; z-index: 5500; pointer-events: none; top: 0px; right: 0px;"
    />
    <span
      id="chakra-toast-manager-bottom-left"
      style="position: fixed; z-index: 5500; pointer-events: none; bottom: 0px; left: 0px;"
    />
    <span
      id="chakra-toast-manager-bottom"
      style="position: fixed; z-index: 5500; pointer-events: none; margin: 0px auto; text-align: center; bottom: 0px; right: 0px; left: 0px;"
    />
    <span
      id="chakra-toast-manager-bottom-right"
      style="position: fixed; z-index: 5500; pointer-events: none; bottom: 0px; right: 0px;"
    />
  </div>
  <div>
    <div
      data-testid="container"
    >
      <div
        class="chakra-portal"
      >
        <h1
          data-testid="heading"
        >
          Hello world
        </h1>
      </div>
    </div>
    <div
      class="chakra-portal-manager"
    />
  </div>
</body>
`;

exports[`should render nested portal 1`] = `
<body
  class="chakra-ui-light"
>
  <div
    id="chakra-toast-portal"
  >
    <span
      id="chakra-toast-manager-top"
      style="position: fixed; z-index: 5500; pointer-events: none; margin: 0px auto; text-align: center; top: 0px; right: 0px; left: 0px;"
    />
    <span
      id="chakra-toast-manager-top-left"
      style="position: fixed; z-index: 5500; pointer-events: none; top: 0px; left: 0px;"
    />
    <span
      id="chakra-toast-manager-top-right"
      style="position: fixed; z-index: 5500; pointer-events: none; top: 0px; right: 0px;"
    />
    <span
      id="chakra-toast-manager-bottom-left"
      style="position: fixed; z-index: 5500; pointer-events: none; bottom: 0px; left: 0px;"
    />
    <span
      id="chakra-toast-manager-bottom"
      style="position: fixed; z-index: 5500; pointer-events: none; margin: 0px auto; text-align: center; bottom: 0px; right: 0px; left: 0px;"
    />
    <span
      id="chakra-toast-manager-bottom-right"
      style="position: fixed; z-index: 5500; pointer-events: none; bottom: 0px; right: 0px;"
    />
  </div>
  <div>
    <div
      class="chakra-portal-manager"
    >
      <div
        class="chakra-portal"
      >
        This is a portal.
        <div
          class="chakra-portal"
        >
          This is a nested portal
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`should render portal 1`] = `
<body
  class="chakra-ui-light"
>
  <div
    id="chakra-toast-portal"
  >
    <span
      id="chakra-toast-manager-top"
      style="position: fixed; z-index: 5500; pointer-events: none; margin: 0px auto; text-align: center; top: 0px; right: 0px; left: 0px;"
    />
    <span
      id="chakra-toast-manager-top-left"
      style="position: fixed; z-index: 5500; pointer-events: none; top: 0px; left: 0px;"
    />
    <span
      id="chakra-toast-manager-top-right"
      style="position: fixed; z-index: 5500; pointer-events: none; top: 0px; right: 0px;"
    />
    <span
      id="chakra-toast-manager-bottom-left"
      style="position: fixed; z-index: 5500; pointer-events: none; bottom: 0px; left: 0px;"
    />
    <span
      id="chakra-toast-manager-bottom"
      style="position: fixed; z-index: 5500; pointer-events: none; margin: 0px auto; text-align: center; bottom: 0px; right: 0px; left: 0px;"
    />
    <span
      id="chakra-toast-manager-bottom-right"
      style="position: fixed; z-index: 5500; pointer-events: none; bottom: 0px; right: 0px;"
    />
  </div>
  <div>
    <div
      class="chakra-portal-manager"
    >
      <div
        class="chakra-portal"
      >
        This is a portal
      </div>
    </div>
  </div>
</body>
`;
